endstate_correction/smc.py

```diff
@@ -0,0 +1,146 @@
+# Perform SMC sampling for endstate correction
+
+import numpy as np
+from mdtraj import Trajectory
+from openmm import unit
+from openmm.app import Simulation
+from scipy.special import logsumexp
+from tqdm import tqdm
+
+from endstate_correction.constant import kBT, temperature
+
+
+class SMC:
+    def __init__(
+        self,
+        sim: Simulation,
+        samples: Trajectory,
+    ) -> None:
+        """Initialize the SMC class
+
+        Args:
+            sim (Simulation): simulation instance
+            samples (Trajectory): samples from which the starting points fo the SMC switching simulation are drawn
+        """
+
+        self.sim = sim
+        self.samples = samples
+        self.logZ = 0.0
+
+    @staticmethod
+    def _calculate_potential_E_for_particles(
+        lamb: float, walkers, sim: Simulation
+    ) -> np.ndarray:
+        # set lambda parameter
+        sim.context.setParameter("lambda_interpolate", lamb)
+
+        u_intermediate = np.zeros(len(walkers))
+        # for each particle calculate the intermediate potential in kBT
+        for p_idx, p in enumerate(walkers):
+            sim.context.setPositions(p)
+            e_pot = sim.context.getState(getEnergy=True).getPotentialEnergy() / kBT
+            u_intermediate[p_idx] = e_pot
+
+        return u_intermediate
+
+    @staticmethod
+    def _propagate_particles(walkers, sim: Simulation):
```

Comment:
Would be good to standardise on naming - either particles or walkers, otherwise it's confusing

Reply:
absolutley! good point!